src/Umbraco.Infrastructure/Services/Implement/PackagingService.cs

```diff
@@ -95,10 +106,60 @@ public void DeleteCreatedPackage(int id, int userId = Cms.Core.Constants.Securit
 
         public string ExportCreatedPackage(PackageDefinition definition) => _createdPackages.ExportPackage(definition);
 
+        public InstalledPackage GetInstalledPackageByName(string packageName)
+            => GetAllInstalledPackages().Where(x => x.PackageName.InvariantEquals(packageName)).FirstOrDefault();
+
+        public IEnumerable<InstalledPackage> GetAllInstalledPackages()
+        {
+            IReadOnlyDictionary<string, string> keyValues = _keyValueService.FindByKeyPrefix(Constants.Conventions.Migrations.KeyValuePrefix);
+
+            var installedPackages = new Dictionary<string, InstalledPackage>();
+
+            // Collect the package from the package migration plans
+            foreach(PackageMigrationPlan plan in _packageMigrationPlans)
+            {
+                if (!installedPackages.TryGetValue(plan.PackageName, out InstalledPackage installedPackage))
+                {
+                    installedPackage = new InstalledPackage
+                    {
+                        PackageName = plan.PackageName
+                    };
+                    installedPackages.Add(plan.PackageName, installedPackage);                    
+                }
+
+                var currentPlans = installedPackage.PackageMigrationPlans.ToList();
+                keyValues.TryGetValue(Constants.Conventions.Migrations.KeyValuePrefix + plan.PackageName, out var currentState);
+                currentPlans.Add(new InstalledPackageMigrationPlans
+                {
+                    CurrentMigrationId = currentState,
+                    FinalMigrationId = plan.FinalState
+                });
+
+                installedPackage.PackageMigrationPlans = currentPlans;
+            }
+
+            // Collect and merge the packages from the manifests
+            foreach(PackageManifest package in _manifestParser.GetManifests())
+            {
+                if (!installedPackages.TryGetValue(package.PackageName, out InstalledPackage installedPackage))
+                {
+                    installedPackage = new InstalledPackage
+                    {
+                        PackageName = package.PackageName
+                    };
+                    installedPackages.Add(package.PackageName, installedPackage);
+                }
+
+                installedPackage.PackageView = package.PackageView;
+            }
```

Comment:
If the packagename is not required, we could just use the foldername inside app_plugins. 
So If I have the following :

* `~\App_Plugins\My.Package1\package.manifest`  with the following content
```
{
  "name": "My Package",
  "packageView" :  "~/App_Plugins/My.Package1/my-package.html"
}
```
* `~\App_Plugins\My.Package2\package.manifest`  with the following content
```
{
  "packageView" :  "~/App_Plugins/My.Package2/my-package.html"
}
```
* `~\App_Plugins\My.Package3\package.manifest`  with the following content
```
{
  "name": "My Package 3",
}
```

This would list:
- "My Package" with package options (Name from package.manifest)
- "My.Package2" with package options (Name from folder name)
- "My Package 3" If and only if there exist package migrations (Name from package.manifest)

```suggestion
            // Collect and merge the packages from the manifests
            foreach(PackageManifest package in _manifestParser.GetManifests())
            {
                var packageName = package.PackageName ?? Path.GetFileName(Path.GetDirectoryName(package.Source));
                if (packageName is null)
                {
                    continue;
                }

                if (!installedPackages.TryGetValue(packageName, out InstalledPackage installedPackage))
                {
                    installedPackage = new InstalledPackage
                    {
                        PackageName = packageName
                    };
                    installedPackages.Add(packageName, installedPackage);
                }

                installedPackage.PackageView = package.PackageView;
            }
```

Reply:
But IMO the package name from the directory name should be done at the package manifest parser. I'll update.